### ISSUE TITLE
Gatsby doesn't provide stub location prop to views during build

### ISSUE THREAD

**Issue by bvaughn** [gatsbyjs/gatsby#2180]
# Steps to reproduce

First, create a new site (`gatsby new test`). Then add the following layout and page components:

## Layout
```jsx
import PropTypes from 'prop-types'
import React from 'react'

const Layout = ({ children, location }) => {
  console.log('<Layout> location:', location);
  return (
    <div>
      {children()}
    </div>
  );
}

Layout.propTypes = {
  children: PropTypes.func.isRequired,
  location: PropTypes.object.isRequired,
};

export default Layout;
```
## Page
```jsx
import PropTypes from 'prop-types'
import React from 'react'

const Page = ({ location }) => {
  console.log('<Page> location:', location);
  return (
    <div>This is a page</div>
  );
}

Page.propTypes = {
  location: PropTypes.object.isRequired,
};

export default Page;
```

Now run `gatsby develop` and look in the console. You'll see:
```
<Layout> location: Object
<Page> location: Object
```

Next run `gatsby build` and look in the terminal output:
```
<Layout> location: { pathname: '/', search: '', hash: '' }
<Page> location: undefined
```

The layout gets a mock location object but the page doesn't. The page is created by invoking the child function passed to layout, so it looks like you can manually pass that prop along (via `children({ location })`) but this, unfortunately, breaks development mode<sup>1</sup> (and also seems unintuitive since there would be different default behavior between the two).

<sup>1</sup> Passing the Gatsby `location` object during development mode causes a react-router error:
```
Uncaught Error: unknown browser :(
    at mapStackTrace (commons.js:106603)
    at RedBoxError.mapError (commons.js:97076)
    at RedBoxError.componentDidMount (commons.js:96996)
    at commons.js:77050
    at measureLifeCyclePerf (commons.js:76861)
    at commons.js:77049
    at CallbackQueue.notifyAll (commons.js:73261)
    at ReactReconcileTransaction.close (commons.js:83407)
    at ReactReconcileTransaction.closeAll (commons.js:86371)
    at ReactReconcileTransaction.perform (commons.js:86318)
```

**Comment by bvaughn** [comment]
At the time, this hack works, but it's pretty gross:
```js
const childrenParams = typeof location.constructor === 'function'
  ? undefined
  : {location};

children(childrenParams);
```

**Comment by bvaughn** [comment]
I think this hack is a viable workaround (but it's pretty gross) 😁 :

```js
const {children, location} = this.props;
const renderedChildren = typeof window === 'undefined'
      ? children({location})
      : children();
```

**Comment by angeloocana** [comment]
Hi guys!
It was fixed but when I upgrade from gatsby@1.9.36 to gatsby@1.9.128 the error starts to happen again. =(

Should I open a new issue?

**Comment by KyleAMathews** [comment]
Yeah a new issue with a reproduction would be ideal.
